src/nameresolver.nim

```diff
@@ -8,63 +8,73 @@
 
 ## Resolves package names and turn them to URLs.
 
-import std / [os, unicode, strutils, osproc, sequtils, options]
+import std / [os, unicode, strutils, osproc, options]
 import context, osutils, packagesjson, gitops
 
 export options
 
+proc retryUrl(cmd, urlstr: string; c: var AtlasContext; repo: PackageRepo;
+              tryBeforeSleep = true): bool =
+  ## Retries a url-based command `cmd` with an increasing delay.
+  ## Performs an initial request when `tryBeforeSLeep` is `true`.
+  const Pauses = [0, 1000, 2000, 3000, 4000, 6000]
+  let firstPause = if tryBeforeSleep: 0 else: 1
+  for i, pause in Pauses[firstPause..^1].pairs():
+    if i > 0: infoNow c, repo, "Retrying remote URL: " & urlstr
+    os.sleep(pause)
+    if execCmdEx(cmd)[1] == QuitSuccess: return true
+  false
+
 proc cloneUrlImpl(c: var AtlasContext,
                   url: PackageUrl,
                   dest: string;
-                  cloneUsingHttps: bool):
-                (CloneStatus, string) =
+                  cloneUsingHttps: bool): (CloneStatus, string) =
   ## Returns an error message on error or else "".
   assert not dest.contains("://")
-  result = (OtherError, "")
-  var modUrl = url
-  if url.scheme == "git" and cloneUsingHttps:
-    modUrl.scheme = "https"
 
-  if url.scheme == "git":
-    modUrl.scheme = "" # git doesn't recognize git://
+  let (repo, url, urlstr, isGitHub) = block:
+    var modurl = url
+    if url.scheme == "git":
+      modurl.scheme = if cloneusinghttps:
+          "https"
+        else:
+          "" # git doesn't recognize git://
+    let isGitHub = modurl.hostname == "github.com"
+    if isGitHub and modurl.path.endswith("/"):
+      # github + https + trailing url slash causes a
+      # checkout/ls-remote to fail with repository not found
+      modurl.path = modurl.path[0 .. ^2]
+    (toRepo($modurl), modurl, $modurl, isGitHub)
```

Comment:
toRepo returns a PackageRepo which is just a URL if I recall correctly. So you could skip the toRepo bit